src/commands/database-remove.js

```diff
@@ -38,32 +40,199 @@ module.exports = new Command("database:remove <path>")
+              }
+              var data = {};
+              try {
+                data = JSON.parse(body);
+              } catch (e) {
+                return reject(
+                  new FirebaseError("Malformed JSON response in shallow get ", {
+                    exit: 2,
+                    original: e,
+                  })
+                );
+              }
+              if (data) {
+                var keyList = Object.keys(data);
+                return resolve(keyList);
+              } else {
+                return resolve([]);
+              }
+            });
+          });
+        });
+      }
+
+      var deleteQueue = [path];
+      var failures = [];
+      var openChunkDeleteJob = 0;
+
+      function dfsLoop() {
+        if (failures.length !== 0) {
+          return true;
+        }
+        if (deleteQueue.length === 0) {
+          return openChunkDeleteJob === 0;
+        }
+        if (openChunkDeleteJob < options.concurrent) {
+          chunckedDelete(deleteQueue.pop());
+        }
+        return false;
+      }
+
+      function chunckedDelete(path) {
+        openChunkDeleteJob += 1;
+        return prefetchTest(path)
+          .then(function(test) {
+            if (test === "small") {
+              return deletePath(path);
+            } else if (test === "larger") {
+              return listPath(path).then(function(pathList) {
+                if (pathList) {
+                  deleteQueue.push(path);
+                  for (var i = 0; i < pathList.length; i++) {
+                    var subPath = path + (path === "/" ? "" : "/") + pathList[i];
+                    deleteQueue.push(subPath);
+                  }
+                }
+                return Promise.resolve();
+              });
+            } else if (test === "empty") {
+              return Promise.resolve();
+            } else {
+              logger.error("unexpected prefetch test result: " + test);
               return reject(
-                new FirebaseError("Unexpected error while removing data", {
+                new FirebaseError("unexpected prefetch test result: " + test, {
                   exit: 2,
+                  original: e,
                 })
               );
-            } else if (res.statusCode >= 400) {
-              return reject(responseToError(res, body));
             }
-
-            utils.logSuccess("Data removed successfully");
-            return resolve();
+          })
+          .then(function() {
+            openChunkDeleteJob -= 1;
+          })
+          .catch(function(error) {
+            openChunkDeleteJob -= 1;
+            failures.push(error);
           });
-        });
+      }
+
+      return new Promise(function(resolve, reject) {
+        var intervalId = setInterval(function() {
+          if (dfsLoop()) {
+            clearInterval(intervalId);
+
+            if (failures.length == 0) {
+              return resolve();
+            } else if (failures.length == 0) {
```

Comment:
This check (`failures.length == 0`) is the same as above. Is this supposed to be `> 0`?